### PR TITLE
Annazhang include hidden questions in application pdf

### PR DESCRIPTION
**Pull request by azhangggggg**
### Description

Please explain the changes you made here.

## Release notes

The title of the pull request will be used as the default release notes description. If more detail is needed to communicate to partners the scope of the PR's changes, use this release notes section.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [ ] Added the correct label: < feature | enhancement | bug | dependencies | infrastructure | ignore-for-release | database >
- [ ] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers).
- [ ] Created unit and/or browser tests which fail without the change (if possible)
- [ ] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [ ] Extended the README / documentation, if necessary

#### User visible changes

- [ ] Followed steps to [internationalize new strings](https://github.com/civiform/civiform/wiki/Internationalization-%28i18n%29#internationalization-for-application-strings)
- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [ ] Tested on mobile view. See [mobile device mode](https://developer.chrome.com/docs/devtools/device-mode/)
- [ ] Manually tested at 200% size
- [ ] Manually evaluated tab order

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/helper/feature-flags.conf`
- [ ] Conditioned new functionality on a [FeatureFlag](https://github.com/civiform/civiform/wiki/Feature-Flags)
- [ ] Wrote browser tests with the feature flag off and on, etc.

### Instructions for manual testing

If instructions are needed for manual testing by reviewers, include them here.

### Issue(s) this completes

Fixes #<issue_number>; Fixes #<issue_number>...

### CODE REVIEW

File: server/app/services/export/PdfExporter.java [hunk 1]
```diff
@@ -184,6 +192,26 @@ private byte[] buildPDF(
           document.add(eligibility);
         }
       }
+      if (answersOnlyHidden.size() > 0) {
```
Comment (gwendolyngoetz) [inline server/app/services/export/PdfExporter.java:195]:
```suggestion
      if (!answersOnlyHidden.isEmpty()) {
```

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java [hunk 2]
```diff
@@ -363,7 +418,8 @@ private ImmutableList<Block> getBlocks(
               blockDefinition,
               applicantData,
               maybeRepeatedEntity);
-      if (includeBlockIfTrue.test(block)) {
+      if ((includeBlockIfTrue.test(block) && !includeHidden)
+          || (!includeBlockIfTrue.test(block) && includeHidden)) {
```
Comment (gwendolyngoetz) [inline server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:422]: (blocking) This is a getting a little hard to follow. Can you move the check into its own method and include comments on what it's doing.

Something like

```Java
/**
 *
 */
private Boolean includeBlockInList(Predicate<Block> includeBlockIfTrue, Boolean includeHidden) {
   return (includeBlockIfTrue.test(block) && !includeHidden) || (!includeBlockIfTrue.test(block) && includeHidden);
}
```

or maybe more verbose for future readers

```Java
/**
 *
 */
private Boolean includeBlockInList(Predicate<Block> includeBlockIfTrue, Boolean includeHidden) {
  // What case does this cover?
  if (includeBlockIfTrue.test(block) && !includeHidden) {
    return true;
  }

  // What case does this cover?
  if (!includeBlockIfTrue.test(block) && includeHidden) {
    return true;
  }

  // What case does this cover?
   return false;
}
```

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java [hunk 3]
```diff
@@ -275,6 +325,9 @@ public ImmutableList<AnswerData> getSummaryDataOnlyActive() {
         String questionText = applicantQuestion.getQuestionText();
         String questionTextForScreenReader = applicantQuestion.getQuestionTextForScreenReader();
         String answerText = applicantQuestion.getQuestion().getAnswerString();
+        if (hiddenBlocks) {
+          answerText = "N/A";
```
Comment (gwendolyngoetz) [inline server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:329]: (question) Does this need to be localized?

**Review by gwendolyngoetz — CHANGES_REQUESTED** [review]

File: server/app/services/applicant/ReadOnlyApplicantProgramService.java [hunk 4]
```diff
@@ -117,8 +129,11 @@ public interface ReadOnlyApplicantProgramService {
   /** Returns if the block has an eligibility predicate. */
   boolean blockHasEligibilityPredicate(String blockId);
 
-  /** Returns if the block eligibility criteria are met. */
-  boolean isBlockEligible(String blockId);
+  /** Returns if the hidden block eligibility criteria are met. */
+  boolean isHiddenBlockEligible(String blockId);
```
Comment (avaleske) [inline server/app/services/applicant/ReadOnlyApplicantProgramService.java:133]: *question*: Is this used anywhere outside of a test? If not, do we need it?

File: server/app/services/applicant/ReadOnlyApplicantProgramService.java [hunk 5]
```diff
@@ -65,6 +71,9 @@ public interface ReadOnlyApplicantProgramService {
    */
   ImmutableList<ApplicantQuestion> getIneligibleQuestions();
 
+  /** Get the hidden block with the given block ID if there is one. It is emptry if there isn't. */
```
Comment (avaleske) [inline server/app/services/applicant/ReadOnlyApplicantProgramService.java:74]: *nit* (typo): empty

File: server/app/services/applicant/ReadOnlyApplicantProgramService.java [hunk 6]
```diff
@@ -35,6 +35,12 @@ public interface ReadOnlyApplicantProgramService {
    */
   ImmutableList<Block> getAllActiveBlocks();
 
+  /**
+   * Get the {@link Block}s for this program and applicant. This includes blocks that are hidden
```
Comment (avaleske) [inline server/app/services/applicant/ReadOnlyApplicantProgramService.java:39]: *suggestion*: This should say it's "only" hidden blocks, not that it "includes" hidden blocks, right?

File: server/app/services/applicant/ReadOnlyApplicantProgramService.java [hunk 7]
```diff
@@ -92,6 +101,9 @@ public interface ReadOnlyApplicantProgramService {
    */
   Optional<Block> getFirstIncompleteBlockExcludingStatic();
 
+  /** Returns summary data for each question in this application. */
```
Comment (avaleske) [inline server/app/services/applicant/ReadOnlyApplicantProgramService.java:104]: *suggestion*: This is only summary data for the hidden blocks, right, not for "each question"? This comment should specify that it only returns summary data for the hidden blocks.

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java [hunk 8]
```diff
@@ -152,10 +164,18 @@ public boolean isBlockEligible(String blockId) {
...
+      allActiveBlockList = getBlocks(this::showBlock, /* includeHidden= */ false);
     }
-    return allBlockList;
+    return allActiveBlockList;
+  }
+
+  @Override
+  public ImmutableList<Block> getAllHiddenBlocks() {
+    if (allHiddenBlockList == null) {
+      allHiddenBlockList = getBlocks(this::showBlock, /* includeHidden= */ true);
```
Comment (avaleske) [inline server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:176]: *question*: Does `includeHidden` mean that it returns active _and_ hidden blocks, or only hidden blocks?

If it's active and hidden, then calls above this might get active blocks unexpectedly since `getAllHiddenBlocks()` implies you're only getting the hidden ones.

*suggestion*:
If it's only hidden blocks, then we should rename the parameter to `onlyHidden` so it's more clear.

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java (hunk 2 above)
Comment (avaleske) [reply server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:422]: +1.

Another way to simplify this might be to remove the check for `includeHidden` and flip the logic in the includeBlockIfTrue predicate. so then line 168 might look like
`allActiveBlockList = getBlocks((block) -> showBlock(block));`

but line 178 might look like
`allHiddenBlockList = getBlocks(block -> !showBlock(block));`

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java (hunk 8 above)
Comment (avaleske) [reply server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:176]: See also the note below about combining `includeHidden` with the `includeIfTrue` predicate.

File: server/app/services/export/PdfExporter.java [hunk 9]
```diff
@@ -184,6 +192,26 @@ private byte[] buildPDF(
...
+        document.add(Chunk.NEWLINE);
+        Paragraph hiddenText =
+            new Paragraph(
+                "Hidden Questions : ", FontFactory.getFont(FontFactory.HELVETICA_BOLD, 12));
```
Comment (avaleske) [inline server/app/services/export/PdfExporter.java:199]: _suggestion_: If there are other section headings in the doc, can we match their font size? It's currently the same font size as a question, which makes it hard to skim.

File: server/test/services/export/PdfExporterTest.java [hunk 10]
```diff
@@ -150,6 +153,30 @@ public void validatePDFExport_OptionalFileUploadWithoutFile()
     }
   }
 
+  @Test
+  public void validatePDFExport_visibility() throws IOException, DocumentException {
```
Comment (avaleske) [inline server/test/services/export/PdfExporterTest.java:157]: *suggestion*: it should be clear from the test name what failed, if the test fails. so something like `validatePdfExport_hiddenQuestionAreIncluded` might be clearer

*also nit, optional*: Three letter acronyms should only have the first letter capitalized, but then that wouldn't match the other method names in this file. Up to you which way you want to go.

**Review by avaleske — CHANGES_REQUESTED** [review]
Thanks for these changes!

I have a few suggestions but overall it looks good.

Also, are there tests for the controllers? As it is, we wouldn't catch a bug that accidentally started including the hidden questions in the user's PDF export

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java (hunk 2 above)
Comment (azhangggggg) [reply server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java]: Changes made. Thank you both for the suggestions!

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java (hunk 3 above)
Comment (azhangggggg) [reply server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java]: Not this doesn't. changes made.

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java [hunk 11]
```diff
@@ -165,15 +161,16 @@ private boolean isBlockEligible(Block block) {
...
+      allActiveBlockList = getBlocks((block) -> showBlock(block));
     }
     return allActiveBlockList;
   }
 
   @Override
   public ImmutableList<Block> getAllHiddenBlocks() {
     if (allHiddenBlockList == null) {
-      allHiddenBlockList = getBlocks(this::showBlock, /* includeHidden= */ true);
+      //
```
Comment (avaleske) [inline server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java:172]: nit (blocking): can we remove this commented line?

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java (hunk 3 above)
Comment (avaleske) [reply server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java]: I think @gwendolyngoetz was asking if the "N/A" text needs to be translated into the user's language, not if it should not be a local variable.

The PDF probably should be translated, but other text in it is not so we shouldn't block on it. For example, the labels on each field: https://github.com/civiform/civiform/blob/cce0552baec4746f118423a64f6c37c09c7f2c5d/server/app/services/export/PdfExporter.java#L113C16-L113C31

**Review by avaleske — APPROVED** [review]
Thanks for these changes!

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java (hunk 11 above)
Comment (azhangggggg) [reply server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java]: yea for sure. deleted.

File: server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java (hunk 3 above)
Comment (azhangggggg) [reply server/app/services/applicant/ReadOnlyApplicantProgramServiceImpl.java]: Thanks for providing more context!
